File: libs/core/wrap.cpp
#include <Adafruit_CircuitPlayground.h>

enum Button
{
    //% block=left
    Left = 1, //CPLAY_LEFTBUTTON,
    //% block=right
    Right = 2 //CPLAY_RIGHTBUTTON
};

enum MotionAxis
{
    X = 0,
    Y = 1,
    Z = 2  
};

enum Animation
{
    Rainbow,
    Sparkle
};

enum Drawing
{
    DrawRainbow
};

enum Note
{
    //% blockIdentity=playground.noteFrequency
    C = 262,
    //% block=C#
    //% blockIdentity=playground.noteFrequency
    CSharp = 277,
    //% blockIdentity=playground.noteFrequency
    D = 294,
    //% blockIdentity=playground.noteFrequency
    Eb = 311,
    //% blockIdentity=playground.noteFrequency
    E = 330,
    //% blockIdentity=playground.noteFrequency
    F = 349,
    //% block=F#
    //% blockIdentity=playground.noteFrequency
    FSharp = 370,
    //% blockIdentity=playground.noteFrequency
    G = 392,
    //% block=G#
    //% blockIdentity=playground.noteFrequency
    GSharp = 415,
    //% blockIdentity=playground.noteFrequency
    A = 440,
    //% blockIdentity=playground.noteFrequency
    Bb = 466,
    //% blockIdentity=playground.noteFrequency
    B = 494
};

enum CapacityPin
{
    //% block="RX #0"
    P0 = 0,
    //% block="TX #1"
    P1 = 1,
    //% block=SDA #2
    P2 = 2,
    //% block=SCL #3
    P3 = 3,
    //% block=#6
    P6 = 6,
    //% block=#9
    P9 = 9,
    //% block=#10
    P10 = 10,
    //% block=#12
    P12 = 12
};

/**
 * Well known colors
 */
enum Color {
    //% block=red blockIdentity=light.color
    Red = 0xFF0000,
    //% block=orange blockIdentity=light.color
    Orange = 0xFFA500,
    //% block=yellow blockIdentity=light.color
    Yellow = 0xFFFF00,
    //% block=green blockIdentity=light.color
    Green = 0x00FF00,
    //% block=blue blockIdentity=light.color
    Blue = 0x0000FF,
    //% block=indigo blockIdentity=light.color
    Indigo = 0x4b0082,
    //% block=violet blockIdentity=light.color
    Violet = 0x8a2be2,
    //% block=purple blockIdentity=light.color
    Purple = 0xFF00FF,
    //% block=white blockIdentity=light.color
    White = 0xFFFFFF
};

enum TemperatureUnit {
    Celsius,
    Fahrenheit
};


/**
* Sensors
*/
//% color=#FE49C9 weight=99
namespace sensors
{
/**
* Gets a value indicating if the left button is pressed.
*/
//% blockId="leftButton" block="%b|button pressed?" weight=40
//% weight=85
boolean button(Button b)
{
    if (b == Button::Left) return CircuitPlayground.leftButton();
    else return CircuitPlayground.rightButton();
}


/**
* Reads the number of taps
*/
//% blockId="getAccelTap" block="taps"
//% weight=84
uint8_t taps() {
    return CircuitPlayground.getAccelTap();
}

/**
* Reads the light level between 0 and 1023.
*/
//% blockId="lightSensor" block="light level"
//% weight=80
uint16_t lightLevel()
{
    return CircuitPlayground.lightSensor();
}

/**
* Reads the accelerometer's Motion
* @param axis the axis of rotation. X is aligned with the buttons, Y going accross the buttons, Z perpendicular to the board.
*/
//% blockId="motion" block="motion %axis"
//% weight=82
int motion(MotionAxis axis) {
    switch(axis) {
        case MotionAxis::Y: return CircuitPlayground.motionY();
        case MotionAxis::Z: return CircuitPlayground.motionZ();
        default: return CircuitPlayground.motionX();
    }
}

/**
* Reads the sound level between 0 and 1023.
*/
//% blockId="soundSensor" block="sound"
//% weight=60
uint16_t sound()
{
    return CircuitPlayground.soundSensor();
}

/**
* Gets a value indicating if the slide switched is on.
*/
//% blockId="slideSwitch" block="slide switch"
//% weight=55
boolean slideSwitch()
{
    return CircuitPlayground.slideSwitch();
}

/**
* Reads the capacitiy of a specific pin
* @param pin the number of the pin
* @param samples
*/
//% blockId="readCap" block="sense touch at pin %pin"
//% weight=50
uint16_t readCap(CapacityPin pin, uint16_t samples = 10)
{
    return CircuitPlayground.readCap((uint8_t)pin, samples);
}

void setAccelRange(lis3dh_range_t range) {
    CircuitPlayground.setAccelRange(range);
}

void setAccelTap(uint8_t c, uint8_t clickthresh) {
    CircuitPlayground.setAccelTap(c, clickthresh);
}

/**
* Reads the temperature.
*/
//% blockId="temperatur" block="temperature %unit"
int temperature(TemperatureUnit unit) {
    if (unit == TemperatureUnit::Celsius)
        return CircuitPlayground.temperature();
    else return CircuitPlayground.temperatureF();
    //TODO: should return float
}

}


/**
* Functions for music / audio
*/
//% color=#FF7D7D weight=75
namespace music
{

/**
* Gets the frequency of a note.
* @param note the note name
*/
//% weight=50
//% blockId=noteFrequency block="%note"
//% shim=TD_ID
uint16_t noteFrequency(Note note)
{
    return note;
}

/**
* Plays a given tone
* @param frequency pitch of tone in Hz
* @param time duration of tone in ms.
*/
//% async blockId="playTone" block="play tone at|freq (Hz) %frequency=noteFrequency|for (ms) %time"
void playTone(uint16_t frequency, uint16_t time = 250)
{
    CircuitPlayground.playTone(frequency, time);
}

}

/**
* Functions to manipulate neopixels
*/
//% color=#00a7e9 weight=50
namespace light
{

/**
* Show a preset drawing. eg: Rainbow
*/
//% blockId="showDrawing" block="show %drawing"
//% weight=95
void showDrawing(Drawing drawing)
{

}

/**
* Rotate the pixels forward.
* @param offset number of pixels to rotate forward, eg: 1
*/
//% blockId="rotate" block="rotate pixels by %offset"
//% weight=95
void rotate(uint8_t offset = 1)
{

}

/**
* Just turn on/off the red #13 LED
* @param on a value to turn on/off the LED, eg: true
*/
//% weight=90
//% blockId="redled" block="red led %on"
void redLED(boolean on)
{
    CircuitPlayground.redLED(on);
}

/**
* Sets the RGB color on a pixel
*/
//% weight=85 blockGap=8
//% blockId="setPixelColor" block="set pixel %p|to color %c=pixelcolor"
void setPixelColor(uint8_t p, uint32_t c)
{
    CircuitPlayground.setPixelColor(p, c);
}

/**
* Sets the RGB color on a pixel
*/
//% weight=85 blockGap=8
//% blockId="setPixelColorRgb" block="set pixel %p|to color %c=pixelcolor"
void setPixelColorRgb(uint8_t p, uint16_t r, uint16_t g, uint16_t b)
{
    CircuitPlayground.setPixelColor(p, r, g, b);
}

/**
* Clear pixels
*/
//% weight=84 blockGap=8
//% blockId="clearPixels" block="clear pixels"
void clearPixels()
{
    CircuitPlayground.clearPixels();
}

/**
* Sets the neopixel brightness
* @param Desired brightness. eg: 255
*/
//% weight=80
//% blockId="setBrightness" block="set brightness %b"
//% advanced=true
void setBrightness(uint16_t brightness)
{
    CircuitPlayground.setBrightness(brightness);
}

/**
* Color wheel
*/
//% blockId="colorWheel" block="color wheel %x"
//% weight=10 blockGap=8
//% advanced=true
uint32_t colorWheel(uint8_t x)
{
    return CircuitPlayground.colorWheel(x);
}


/**
* Converts RGB channels into a color
*/
//% blockId="rgb" block="red %r|green %g|blue %b"
//% weight=9
//% advanced=true
uint32_t rgb(uint8_t r, uint8_t g, uint8_t b)
{
    return r >> 16 | g >> 8 | r;
}
}

/**
* Functions to manipulate serial
*/
//% color=#00a700 weight=40
//% advanced=true
namespace serial {
/**
* Serial print
*/
//% 
void print(uint16_t code) {
    Serial.println(code);
}
}

File: libs/core/wrap.ts
/**
* Sensors
*/
//% color=#FE49C9 weight=99
namespace sensors {
}

/**
* Functions for playing audio / music
*/
//% color=#FF7D7D weight=75
namespace music {
}

/**
* Functions to manipulate neopixels
*/
//% color=#00a7e9 weight=85
namespace light {
    let NUM_PIXELS: number = 10;

    /**
    * Gets a wellknown color
    */
    //% blockId="pixelcolor" block="%c"
    //% shim=TD_ID blockGap=8
    //% weight=5
    export function color(c: Color): number {
        return c;
    }

    // what's the current high value
    let barGraphHigh = 0;
    // when was the current high value recorded
    let barGraphHighLast = 0;

    /**
     * Displays a vertical bar graph based on the `value` and `high` value.
     * If `high` is 0, the chart gets adjusted automatically.
     * @param value current value to plot
     * @param high maximum value. If 0, maximum value adjusted automatically, eg: 0
     */
    //% weight=20
    //% blockId=bargraph block="bar graph of %value |up to %high" icon="\uf080"
    export function bargraph(value: number, high: number): void {
        /*
        const now = control.millis();
        value = Math.abs(value);

        if (high != 0) barGraphHigh = high;
        else if (value > barGraphHigh || now - barGraphHighLast > 10000) {
            barGraphHigh = value;
            barGraphHighLast = now;
        }

        barGraphHigh = Math.max(barGraphHigh, 1);

        const n = 10;
        const n1 = n - 1;
        const v = (value * n) / barGraphHigh;
        if (v == 0) {
            light.setPixelColor(0, 0x666600);
            for (let i = 1; i < n; ++i)
                light.setPixelColor(i, 0);
        } else {
            for (let i = 0; i < n; ++i) {
                if (i <= v) {
                    const b = i * 255 / n1 & 0xff;
                    light.setPixelColor(i, (b >> 16) | (255 - b));
                }
                else light.setPixelColor(i, 0);
            }
        }
        */
    }

    /**
     * Show a preset animation. eg: Rainbow
     * @param animation the preset animation to show
     */
    //% async blockId="showAnimation" block="show Animation %animation"
    //% weight=85
    export function showAnimation(animation: Animation) {
        switch(animation) {
            case Animation.ClassicRainbow: 
                animateRainbow(350); break;
            case Animation.RainbowCycle: 
                animationRainbowCycle(10); break;
            case Animation.Sparkle:
                animateSparkle(0xff, 0xff, 0xff, 25); break;
            case Animation.ColorWipe:
                animateColorWipe(0x00,0xff,0x00, 50); break;
            case Animation.RunningLights:
                animateRunningLights(0xff,0xff,0x00, 100); break;
            case Animation.TheatreChase:
                animateTheatreChase(0xff,0,0,100); break;
            default:
        }
    }

    // the current pixel in a rainbow animation
    let pixeln = 0;
    function animateRainbow(SpeedDelay: number) {
        light.setPixelColorWheel(pixeln++, pixeln *25);
        if (pixeln == NUM_PIXELS + 1) {
            pixeln = 0;
            clearPixels();
        }
        control.delay(SpeedDelay);
    }

    function animationRainbowCycle(Speed: number) {
        
        // Make an offset based on the current millisecond count scaled by the current speed.
        let offset = control.millis() / Speed;

        // Loop through each pixel and set it to an incremental color wheel value.
        for(let i=0; i < NUM_PIXELS; ++i) {   
            //CircuitPlayground.strip.setPixelColor(i, CircuitPlayground.colorWheel(((i * 256 / 10) + offset) & 255));
            light.setPixelColorWheel(i, (((i * 256 / 10) + offset) & 255));
        }

        // Show all the pixels.
        showStrip();

        //control.delay(1);
        
    }

    function animateSparkle(red: number, green: number, blue: number, SpeedDelay: number) {
        let Pixel = Math.random(NUM_PIXELS);
        setStripPixelColorRgb(Pixel,red,green,blue);
        showStrip();
        control.delay(SpeedDelay);
        setStripPixelColorRgb(Pixel,0,0,0);
        showStrip();
    }

    function animateColorWipe(red: number, green: number, blue: number, SpeedDelay: number) {
        for(let i=0; i<NUM_PIXELS; i++) {
            setPixelColorRgb(i, red, green, blue);
            control.delay(SpeedDelay);
        }
        for(let i=0; i<NUM_PIXELS; i++) {
            setPixelColorRgb(i, 0, 0, 0);
            control.delay(SpeedDelay);
        }
    }

    //TODO: The following animation requires floating pointer support for Math.sin
    let Position = 0;
    function animateRunningLights(red: number, green: number, blue: number, WaveDelay: number) {
        /*
        Position++; // = 0; //Position + Rate;
        for(let i=0; i < NUM_PIXELS; i++) {
            // sine wave, 3 offset waves make a rainbow!
            //float level = sin(i+Position) * 127 + 128;
            //setPixel(i,level,0,0);
            //float level = sin(i+Position) * 127 + 128;
            setPixelColorRgb(i,((Math.sin(i+Position) * 127 + 128)/255)*red,
                    ((Math.sin(i+Position) * 127 + 128)/255)*green,
                    ((Math.sin(i+Position) * 127 + 128)/255)*blue);
        }
        
        showStrip();
        control.delay(WaveDelay);
        
        if (Position == (NUM_PIXELS*2)) {
            Position = 0;
        }*/
    }

    let q = 0;
    function animateTheatreChase(red: number, green: number, blue: number, SpeedDelay: number) {
        for (let i=0; i < NUM_PIXELS; i=i+3) {
            setStripPixelColorRgb(i+q, red, green, blue);    //turn every third pixel on
        }
        showStrip();
        control.delay(SpeedDelay);
        
        for (let i=0; i < NUM_PIXELS; i=i+3) {
            setStripPixelColorRgb(i+q, 0,0,0);        //turn every third pixel off
        }
        q++;
        if (q == 3) {
            q = 0;
        }
    }

    /**
     * Show a preset drawing. eg: Rainbow
     * @param drawing the preset drawing to show
     */
    //% blockId="showDrawing" block="show %drawing"
    //% weight=95
    export function showDrawing(drawing: Drawing)
    {
        switch(drawing) {
            case Drawing.Rainbow: 
                showRainbow(); break;
            case Drawing.Comet: 
                showComet(); break;
            default:
        }
    }

    function showRainbow() {
        for (let i = 0; i < NUM_PIXELS; i++) {
            light.setPixelColorWheel(i, i * 25);
        }
    }

    function showComet() {
        for (let i = 0; i < NUM_PIXELS; i++) {
            setPixelColorRgb(i, 255-(i*25), 0, 0);
        }
    }

    /**
     * Rotate the pixels forward.
     * @param offset number of pixels to rotate forward, eg: 1
     */
    //% blockId="rotate" block="rotate pixels by %offset"
    //% weight=95
    export function rotate(offset: number = 1, reverse?: boolean) {
        /*
        for (let i = 0; i < offset; i++) {
            if(reverse)
                this.neopixels.unshift(this.neopixels.pop());
            else
                this.neopixels.push(this.neopixels.shift());
        }*/
    }

    /*
    function setAll(red: number, green: number, blue: number) {
        for(let i = 0; i < NUM_PIXELS; i++ ) {
            setPixel(i, red, green, blue); 
        }
        showStrip();
    }*/
}
